### PR TITLE
Add scripts to convert v1.x models to v2.0

### PR DESCRIPTION
**Pull request by ZwwWayne**
This PR adds a script to upgrade v1.x models to v2.0 version.
Due to the in-compatibility between V1.x and V2.0, the script does not ensure the v1.x model weights can fully reproduce the same results in the V2.0 codebases.
This PR also meets the requests in https://github.com/open-mmlab/mmdetection/issues/2551.

### CODE REVIEW

File: tools/upgrade_model_version.py [hunk 1]
```diff
@@ -15,17 +99,35 @@ def convert(in_file, out_file):
...
+        new_key = key
+        new_val = val
+        if is_two_stage and is_head(key):
+            new_key = 'roi_head.{}'.format(key)
+
+        # classification
+        m = re.search(r'(conv_cls|retina_cls|rpn_cls|fc_cls).(weight|bias)',
+                      new_key)
+        if m is not None:
+            print(new_key)
```
Comment (xvjiarui) [inline tools/upgrade_model_version.py:115]: Change to `print(f'reorder cls channels of {new_key})`

File: tools/upgrade_model_version.py [hunk 2]
```diff
@@ -15,17 +99,35 @@ def convert(in_file, out_file):
...
+
+        # mask head
+        m = re.search(r'(conv_logits).(weight|bias)', new_key)
         if m is not None:
-            param = m.groups()[1]
-            new_key = key.replace(param, 'conv.{}'.format(param))
-            out_state_dict[new_key] = val
-            continue
+            print(new_key)
+            new_val = truncate_cls_channel(val, num_classes)
```
Comment (xvjiarui) [inline tools/upgrade_model_version.py:128]: print(f'truncate reg channels of {new_key})

**Comment by xvjiarui** [comment]
Add some docs about how to use the script and what performance is expected.

**Comment by charlesyann** [comment]
please check this script whether works well on SSD, it doesn't work for me~~

File: configs/legacy/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# Legacy Configs in MMDetection V1.x
```
Comment (hellock) [inline configs/legacy/README.md:1]: Rename the folder to `legacy_1.x`

File: configs/legacy/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+Configs in this directory implements the legacy configs used by MMDetection V1.x and its model zoos.
```
Comment (hellock) [inline configs/legacy/README.md:3]: implements -> implement

File: configs/legacy/README.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+python -u tools/upgrade_model_version.py ${OLD_MODEL_PATH} ${NEW_MODEL_PATH} --num-classes ${NUM_CLASSES} [--upgrade-retina] [--ssd] [--reg-cls-agnostic]
+
+```
+- OLD_MODEL_PATH: the path to load the model weights in 1.x version.
+- NEW_MODEL_PATH: the path to save the converted model weights in 2.0 version.
+- NUM_CLASSES: number of classes of the original model weights. Usually it is 81 for COCO dataset, 21 for VOC dataset.
+The number of classes in V2.0 models should equal to the number of classes in V1.x models + 1.
+
+Optional arguments:
+- `--upgrade-retina` Option for some RetinaNet models trained in older codebase (lower than V1.0), which will also upgrade the model weights' name.
```
Comment (hellock) [inline configs/legacy/README.md:27]: This is not necessary and we won't support v0.x any more.

File: configs/legacy/README.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+- `--ssd` Option for the SSD Detector, the SSD Detector only need to reorder its classification branch.
+- `--reg-cls-agnostic` Option for Cascade methods whose regression branch is class agnostic.
```
Comment (hellock) [inline configs/legacy/README.md:29]: Config is stored in the meta of checkpoint and we may obtain such information from the config.

File: configs/legacy/README.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+### 2. Use configs with legacy settings
+
+After converting the model weights, create a config that uses the legacy settings, e.g. `LegacyAnchorGenerator`, `LegacyDeltaXYWHBBoxCoder` and `RoIAlign` without aliggment. For models using ResNet Caffe backbones, they also need to change the pretrain name and the corresponding `img_norm_cfg`.
```
Comment (hellock) [inline configs/legacy/README.md:33]: create a config -> checkout to the v1.2 release to find the corresponding config file

File: configs/legacy/README.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+An example is in [`retinanet_r50_caffe_fpn_1x_coco_v1.py`](retinanet_r50_caffe_fpn_1x_coco_v1.py)
+Then use the config to test the model weights. The obtained results should be close to that in V1.x.
```
Comment (hellock) [inline configs/legacy/README.md:35]: If we did not test all models, we may claim:
**For most models,** the obtained results should be close to that in V1.x.

File: configs/legacy/README.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
+# Legacy Configs in MMDetection V1.x
+
+Configs in this directory implements the legacy configs used by MMDetection V1.x and its model zoos.
+
+To help users convert their models from V1.x to MMDetection V2.0, we provide v1.x configs to inference the converted v1.x models.
+Due to the BC-breaking changes in MMDetection V2.0 from MMDetection V1.x, running inference with the same model weights in these two version will produce different results. The difference will cause within 1% AP absolute difference as can be found in the following table.
+
+## Usage
+
+To upgrade the model version, the users need to do the following steps.
```
Comment (hellock) [inline configs/legacy/README.md:10]: We may also give some instructions on how to convert models that are not in the model zoo, e.g., custom detectors.

File: configs/legacy_1.x/README.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+detectors. We provide a scripts `tools/upgrade_model_version.py` to convert the model weights in the V1.x model zoo.
+
+```bash
+python -u tools/upgrade_model_version.py ${OLD_MODEL_PATH} ${NEW_MODEL_PATH} --num-classes ${NUM_CLASSES} [--upgrade-retina] [--ssd] [--reg-cls-agnostic]
+
+```
+- OLD_MODEL_PATH: the path to load the model weights in 1.x version.
+- NEW_MODEL_PATH: the path to save the converted model weights in 2.0 version.
+- NUM_CLASSES: number of classes of the original model weights. Usually it is 81 for COCO dataset, 21 for VOC dataset.
+The number of classes in V2.0 models should equal to the number of classes in V1.x models + 1.
```
Comment (hellock) [inline configs/legacy_1.x/README.md:29]: - should -> should be
- Is this sentence correct?

File: configs/legacy_1.x/README.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+Optional arguments:
+- `--upgrade-retina` Option for some RetinaNet models trained in older codebase (lower than V1.0), which will also upgrade the model weights' name.
+- `--ssd` Option for the SSD Detector, the SSD Detector only need to reorder its classification branch.
```
Comment (hellock) [inline configs/legacy_1.x/README.md:33]: need -> needs

File: docs/COMPATIBILITY.md [hunk 12]
```diff
@@ -50,7 +50,9 @@ thus MMDetection 2.0 do not allow low quality matching by default in R-CNN train
   ```
   [model]_(model setting)_[backbone]_[neck]_(norm setting)_(misc)_(gpu x batch)_[schedule]_[dataset].py,
   ```
-  where the (misc) includes DCN and GCBlock, etc. For example, `reppoints_moment_r101_fpn_gn-neck+head_mstrain_480-960_2x_coco.py` indicates the RepPoints method with moment setting using ResNet101 backbone and FPN with GN in the neck and head, trained by multi-scale augmentation by 2x schedule on COCO dataset.
+  where the (misc) includes DCN and GCBlock, etc. More details are in the [Documentation for config](CONFIG.md)
```
Comment (hellock) [inline docs/COMPATIBILITY.md:53]: - are in -> are illustrated in
- Documentation -> documentation

File: docs/COMPATIBILITY.md [hunk 13]
```diff
@@ -50,7 +50,9 @@ thus MMDetection 2.0 do not allow low quality matching by default in R-CNN train
...
+
+- MMDetection V2.0 uses new ResNet Caffe backbones to reduce warnings when loading pre-trained models. Most of the new backbone's weights are the same as the former ones but do not have `conv.bias`, except that they use a different `img_norm_cfg`. Thus, the new backbone will not cause warning of unexpected keys.
```
Comment (hellock) [inline docs/COMPATIBILITY.md:55]: backbone's -> backbones'

File: docs/COMPATIBILITY.md [hunk 14]
```diff
@@ -67,3 +69,9 @@ This slightly improves both mask AP and bbox AP by ~0.2% absolute.
 - The default setting will not use gradient clipping anymore during training for faster training speed. This does not degrade performance of the most of models. For some models such as RepPoints we keep using gradient clipping to stablize the training process and to obtain better performance.
```
Comment (hellock) [inline docs/COMPATIBILITY.md:69]: will not -> does not

File: tools/upgrade_model_version.py [hunk 15]
```diff
@@ -1,11 +1,117 @@
...
+def is_head(key):
+    valid_head_list = [
+        'bbox_head', 'mask_head', 'semantic_head', 'grid_head', 'mask_iou_head'
+    ]
+
+    return any(key.startswith(h) for h in valid_head_list)
+
+
+def parse_config(config_strings):
+    temp_file = tempfile.NamedTemporaryFile()
```
Comment (hellock) [inline tools/upgrade_model_version.py:19]: It is not deleted after being used.

File: configs/legacy_1.x/README.md [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+- `--reg-cls-agnostic` Option for Cascade methods whose regression branch is class agnostic.
```
Comment (hellock) [inline configs/legacy_1.x/README.md:34]: Give some examples for the usage of this script.

File: configs/legacy_1.x/README.md [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+1. Since the class order in all the detector's classification branch is reordered, all the legacy model weights need to go through the conversion process.
+2. The regression and segmentation head no longer contain the background channel. Weights in these background channels should be removed to fix in the current codebase.
+3. For two-stage detectors, their wegihts need to be upgraded since MMDetection V2.0 refactors all the two-stage detectors with `RoIHead`.
+
+The users can do the same modification as mentioned above for the self-implemented
+detectors. We provide a scripts `tools/upgrade_model_version.py` to convert the model weights in the V1.x model zoo.
+
+```bash
+python -u tools/upgrade_model_version.py ${OLD_MODEL_PATH} ${NEW_MODEL_PATH} --num-classes ${NUM_CLASSES}
```
Comment (hellock) [inline configs/legacy_1.x/README.md:23]: `-u` is unnecessary.

File: tools/upgrade_model_version.py [hunk 18]
```diff
@@ -1,11 +1,117 @@
...
+    config_path = f'{temp_file.name}.py'
+    with open(config_path, 'w') as f:
+        f.write(config_strings)
+
+    config = Config.fromfile(config_path)
+    is_two_stage = True
+    is_ssd = False
+    is_retina = False
+    reg_cls_agnostic = False
+    if 'rpn_head' not in config.model.keys():
```
Comment (hellock) [inline tools/upgrade_model_version.py:29]: `config.model.keys()` -> `config.model`
